src/validate.js

```diff
@@ -0,0 +1,183 @@
+    try {
+      cfg.constants = yaml.safeLoad(fs.readFileSync(fullpath, 'utf-8'));
+    }
+    catch (err) {
+      if (err.code == 'ENOENT') {
+        debug('Constants file does not exist, setting constants to {}');
+        cfg.constants = {};
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  let promises = [];
+  let schemas = [];
+  let ajv = Ajv({useDefaults: true});
+
+  function addSchema (root, name) {
+    let json = null;
+    let data = fs.readFileSync(path.resolve(root, name), 'utf-8');
+    if (/\.ya?ml$/.test(name)) {
+      json = yaml.safeLoad(data);
+    } else {
+      json = JSON.parse(data);
+    }
+
+    let schema = render(json, cfg.constants);
+
+    name = name.replace(/\.ya?ml$/, '.json');
+    let id = urljoin(cfg.baseurl, cfg.prefix, name) + '#';
+    if (!schema.id) {
+      schema.id = id;
+    }
+    if (schema.id !== id) {
+      debug("Bad schema name: %s expected: %s", schema.id, id);
+      throw new Error('Incorrect schemaId specified. It is recommended not to set' +
+          'one at all. It will be set automatically.');
+    }
+
+    try {
+      ajv.addSchema(schema);
+      debug('Loaded schema with id of "%s"', schema.id);
+      schemas.push({name: name, schema: schema});
+    } catch(err) {
+      debug('failed to load schema at %s', path.resolve(root,name));
+      throw err;
+    }
+  }
+
+  function finishSchemaLoading() {
+   debug("finished walking tree of schemas");
+   if (cfg.publish) {
+     debug('Publishing schemas');
+     assert(cfg.aws, "Can't publish without aws credentials.");
+     assert(cfg.prefix, "Can't publish without prefix");
+     assert(cfg.prefix == "" || /.+\/$/.test(cfg.prefix),
+       "prefix must be empty or should end with a slash");
+     let s3Provider = null;
+     if (cfg.s3Provider) {
+       debug('Using user-provided s3 client');
+       s3Provider = cfg.s3Provider;
+     }
+     else {
+       debug('Using default s3 client');
+       s3Provider = new AWS.S3(cfg.aws);
+     }
+     promises = promises.concat(
+       schemas.map( (entry) => {
+         return publishSchema(
+           s3Provider,
+           cfg.bucket,
+           cfg.prefix,
+           entry.name,
+           entry.schema
+         );
+       })
+     );
+   }
+  }
+
+  let walkOptions = {
+    listeners: {
+      name: addSchema,
+      end: finishSchemaLoading,
+    }
+  }
+
+  let walker = walk.walkSync(path.resolve(cfg.folder), walkOptions);
+  await Promise.all(promises);
+
+  return (obj, id) => {
+    id = id.replace(/\.ya?ml$/, '.json');
+    if (!_.endsWith(id, '.json')) {
+      id = id + '.json';
+    }
+    ajv.validate(id, obj);
+    if (ajv.errors) {
+      debug(ajv.errorsText());
+    }
+    return ajv.errors;
+  }
+};
+
+function publishSchema(s3, bucket, prefix, name, content) {
+  return new Promise((accept, reject) => {
+    debug('Publishing schema %s', name);
+    content = JSON.stringify(content, undefined, 4);
+    if (!content) {
```

Comment:
I think `JSON.stringify` will throw an error if it fails... but I could be wrong...